File: problems/arrays_n_hashing/easy/javascript/contains_duplicates.js
// Contains Duplicates
// https://leetcode.com/problems/contains-duplicate/
// Given an integer array nums, return true if any value appears at least twice in the array, and return false if every element is distinct.

// VERSION 1:
// Runtime: 2227 ms
// Memory: 50.2 mb
// Note: 

// -------------------------------------

// var containsDuplicate = (nums) => {
//   // Initialize an empty array to store unique values
//   let uniq = [];
//   // Initialize a boolean variable to keep track of whether duplicates are found
//   let dupes = false;

//   // Iterate through each element in the array
//   nums.forEach((num) => {
//     // If the current element is already in the unique array, set dupes to true
//     if (uniq.indexOf(num) != -1) {
//       dupes = true;
//     } else {
//       // Otherwise, add the current element to the unique array
//       uniq.push(num);
//     }
//   });

//   // Return the value of the dupes variable
//   return dupes;
// };

// ===================================================================== //

// VERSION 2:
// Runtime: 122 ms
// Memory: 50.8 mb
// Note: Uses the Map data structure

// -------------------------------------

// var containsDuplicate = (nums) => {
//   // Create a new Map object
//   const numMap = new Map();

//   // Loop through each number in 'nums'
//   for (let i = 0; i < nums.length; i++) {
//     // If 'numMap' has the number as a key, return 'true' since it is a duplicate
//     if (numMap.has(nums[i])) {
//       return true;
//       // Otherwise, set the number as a key in 'numMap'
//     } else {
//       numMap.set(nums[i]);
//     }
//   }
//   // Return 'false' if no duplicates were found
//   return false;
// };

// ===================================================================== //

// VERSION 3:
// Runtime: 165 ms
// Memory: 50.6 mb
// Note: Uses the Set data structure

// -------------------------------------

// var containsDuplicate = (nums) => {
//   // Use a Set to store the unique values of the input array
//   const uniqueNums = new Set(nums);

//   // Spread the Set back into an array to compare lengths
//   // and check for duplicates
//   const uniqueArr = [...uniqueNums];
//   const hasDuplicates = uniqueArr.length !== nums.length;

//   // Return whether duplicates were found
//   return hasDuplicates;
// };

// ===================================================================== //

console.log(containsDuplicate([1,2,3,1])) //=> true
console.log(containsDuplicate([1,2,3,4])) //=> false
console.log(containsDuplicate([1,1,1,3,3,4,3,2,4,2])) //=> true

File: problems/arrays_n_hashing/easy/ruby/contains_duplicates.rb
# Contains Duplicates
# https://leetcode.com/problems/contains-duplicate/
# Given an integer array nums, return true if any value appears at least twice in the array, and return false if every element is distinct.

# VERSION 1:
# Runtime: 134 ms
# Memory: 219 mb
# Note: Uses built in `uniq` method

# -------------------------------------

# def contains_duplicate(nums)
#   !(nums.length == nums.uniq.length)
# end

# ===================================================================== #

# VERSION 2:
# Runtime: N/A
# Memory: N/A
# Note: Time Limit Exceeded (When array length gets large. `include?` has to search the entire array which is time-consuming)

# -------------------------------------

# def contains_duplicate(nums)
#   uniq_nums = []

#   nums.each do |num|
#     if uniq_nums.include?(num)
#       return true
#     end

#     uniq_nums << num
#   end

#   return false
# end

# ===================================================================== #

# VERSION 3:
# Runtime: 131 ms
# Memory: 224.6 mb
# Note: Hashes are more efficient than arrays

# -------------------------------------

def contains_duplicate(nums)
  nums_hash = {}

  nums.each do |num|
    if nums_hash.has_key?(num)
      return true
    end

    nums_hash[num] = nil
  end

  return false
end

# ===================================================================== #

p contains_duplicate([1,2,3,1]) #=> true
p contains_duplicate([1,2,3,4]) #=> false
p contains_duplicate([1,1,1,3,3,4,3,2,4,2]) #=> true